### ISSUE TITLE
mouse event doesn't make css work

### ISSUE THREAD

**Issue by aisensiy** [n1k0/casperjs#208]
Casperjs can just trigger events write by javascript, but not with css. 

```css
.container .prev {
  display: none;
}

.container:hover .prev {
  display: block;
}
```

With the example above, when i try

```javascript
casper.mouseEvent('mouseover', '.container');
```
The `.prev` elment doesn't show up.

**Comment by gbell12** [comment]
I think something's still wrong.  If I run the example code, the image results indicate that this test:

casper.then(function(){
    this.mouseEvent('mousemove', '#cssPseudo');
    this.captureSelector('bodycss1.png', 'body');
});

Is triggering the yellow square, meaning the #jsHover element is seeing a mouseover instead of the #cssPsuedo element.

**Comment by jefleponot** [comment]
Hi @istr 
this issue is labelled as Bug but it's not exactly a bug but rather a browser limitation.

**Events that are generated by the user agent, either as a result of user interaction, or as a direct result of changes to the DOM, are trusted by the user agent with privileges that are not afforded to events generated by script through the DocumentEvent.createEvent("Event") method, modified using the Event.initEvent() method, or dispatched via the EventTarget.dispatchEvent() method. The isTrusted attribute of trusted events has a value of true, while untrusted events have a isTrusted attribute value of false**

_Most untrusted events should not trigger default actions, with the exception of click or DOMActivate events._

https://www.w3.org/TR/DOM-Level-3-Events/#trusted-events

solutions to workaround this limitation : 
------------------------------------------------
* You have to add a class and remove that on the mouseover/mouseout/mousemove events manually or automaticaly.
* Native Events must be preferred on all the mouseover/mouseout/mousemove events.

To hold my CasperJs compatibility, I use a mixture of the 2 solutions seem to be a good thing.

I had these method on clientutils.js

```
/**
* copyStylesPseudoClasses clone css speudo class as :hover :before :after
* private slot
*
* @param  String  str  The string content speudo class like 'hover, active, target...'
*/
this.copyStylesPseudoClasses = function copyStylesPseudoClasses(pseudoclassname){
"use strict";
 if ( document.getElementById(pseudoclassname+'PseudoClass') ) return;

var NewstylesheetText="",s = document.createElement("style");
for(var i=0; i<document.styleSheets.length; i++) {
    var sheet = document.styleSheets[i];
    for(var j=0; j<sheet.cssRules.length; j++) {
       var cssRule = sheet.cssRules[j];
       var media = "";
       var rules = [cssRule];
       if (cssRule.cssRules) {
          rules = cssRule.cssRules;  
          media="@media "+cssRule.conditionText+" ";
       } 
       for(var k=0; k<rules.length; k++) {
         var t = rules[k].cssText; 
         if (t.toLowerCase().indexOf(':'+pseudoclassname)!==-1){
         NewstylesheetText+=media+rules[k].selectorText.replace(':'+pseudoclassname,'.'+pseudoclassname+'PseudoClass')+
                            '{'+t.substring(t.indexOf('{')+1,t.indexOf('}'))+"}\n";
         }
       }
    }
  }
  s.type = "text/css";
  s.rel = "stylesheet";
  s.id = pseudoclassname+'PseudoClass';
  s.innerHTML = NewstylesheetText;
  document.getElementsByTagName("head")[0].appendChild(s);
};

/**
* addPseudoClassToElement add a speudo class to simulate css hover style and other speudo class
*
* @param  String  pseudoclassname     Type of css speudoClass
* @param  String  selector  A CSS3 selector to the element to click
*/
this.addPseudoClassToElement = function addPseudoClassToElement(pseudoclassname, selector){
  "use strict";
  var elem = this.findOne(selector);
  this.copyStylesPseudoClasses(pseudoclassname);
   while(!!elm && elm!==document.body){
      elm.className += ' '+pseudoclassname+'PseudoClass';
      elm=elm.parentNode;
   }
};

/**
* removePseudoClassToElement remove a speudo class to clean css hover style and other speudo class
*
* @param  String  pseudoclassname     Type of css speudoClass
* @param  String  selector  A CSS3 selector to the element to click
*/
this.removePseudoClassToElement = function removePseudoClassToElement( pseudoclassname, selector) {
  "use strict";
  var elem = this.findOne(selector);
  while(!!elm && elm!==document.body){
    var cn = elm.className;
    var rxp = new RegExp( "\\s?\\b"+pseudoclassname+'PseudoClass'+"\\b", "g" );
    cn = cn.replace( rxp, '' );
    elm.className = cn;
    elm=elm.parentNode;
  }
  var stylesheet = document.getElementById(pseudoclassname+'PseudoClass');
  if (!!stylesheet) stylesheet.parentNode.removeChild(stylesheet);
};
```
this code work with 90% of css pseudo class writing (not so bad)
Are the main contributors agree to propose this kind of workaround ?

Regards
Jef

**Comment by istr** [comment]
> this issue is labelled as Bug but it's not exactly a bug but rather a browser limitation.

@jefleponot 
Yes, it is old and was labelled "bug" before I dug into this project... We could remove the bug flag, but OTOH we know the "browser limitations" of our underlying engines pretty well and already "work around" other (webkit) limitations for older versions. So there is some justification for the "bug" label.

If any of the other maintainers feels this is a "non-bug", feel free to re-label.

**Comment by istr** [comment]
This one is addressed in #1417 at least for some engines.

**Comment by istr** [comment]
Closing by means of merged #1417.
